cis194/week2/tien/homework2/src/LogAnalysis.hs

```diff
@@ -1,26 +1,56 @@
 {-# OPTIONS_GHC -Wall #-}
+{-# LANGUAGE ViewPatterns #-}
 
 module LogAnalysis where
 
 import           Log
 import           Text.Read (readMaybe)
 
+-- Exercise 1 --
+-- parseMessage :: String -> LogMessage
+-- parseMessage logStr =
+--   case words logStr of
+--     ("I":ts:rest) ->
+--       case readMaybe ts :: Maybe Int of
+--         Just i -> LogMessage Info i (unwords rest)
+--         _      -> Unknown logStr
+--     ("W":ts:rest) ->
+--       case readMaybe ts :: Maybe Int of
+--         Just i -> LogMessage Warning i (unwords rest)
+--         _      -> Unknown logStr
+--     ("E":lv:ts:rest) ->
+--       case (readMaybe lv :: Maybe Int, readMaybe ts :: Maybe Int) of
+--         (Just i, Just j) -> LogMessage (Error i) j (unwords rest)
+--         _                -> Unknown logStr
+--     _ -> Unknown logStr
+-- Use ViewPatterns --
 parseMessage :: String -> LogMessage
-parseMessage logStr =
-  case words logStr of
-    ("I":ts:rest) ->
-      case readMaybe ts :: Maybe Int of
-        Just i -> LogMessage Info i (unwords rest)
-        _      -> Unknown logStr
-    ("W":ts:rest) ->
-      case readMaybe ts :: Maybe Int of
-        Just i -> LogMessage Warning i (unwords rest)
-        _      -> Unknown logStr
-    ("E":lv:ts:rest) ->
-      case (readMaybe lv :: Maybe Int, readMaybe ts :: Maybe Int) of
-        (Just i, Just j) -> LogMessage (Error i) j (unwords rest)
-        _                -> Unknown logStr
-    _ -> Unknown logStr
+parseMessage (words -> "I":(readMaybe -> Just ts):msg) =
+  LogMessage Info ts (unwords msg)
+parseMessage (words -> "W":(readMaybe -> Just ts):msg) =
+  LogMessage Warning ts (unwords msg)
+parseMessage (words -> "E":(readMaybe -> Just lvl):(readMaybe -> Just ts):msg) =
+  LogMessage (Error lvl) ts (unwords msg)
+parseMessage logStr = Unknown logStr
 
 parse :: String -> [LogMessage]
 parse = fmap parseMessage . lines
+
+-- Exercise 2 --
+constrcutTreeByTs ts1 ts2 log1 log2 left right
+  | ts1 > ts2 = Node left log2 (insert log1 right)
+  | ts1 < ts2 = Node (insert log1 left) log2 right
```

Comment:
That will result in a non exhaustive pattern matching warning. Maybe you can either use `ts1 <= ts2` or just use `otherwise`. You may turn on `--ghc-options="-Wall -Werror"` in your `stack` command and you will be forced to fix those :)